### ISSUE TITLE
Static error-(404|500).html not honouring https protocol from .env SS_BASE_URL

### ISSUE THREAD

**Issue by axllent** [silverstripe/silverstripe-errorpage#20]
When static error pages are generated from the cli with `dev/build` the `error-404.html` and `error-500.html` files contain a `<base>` url that appears hard-coded to use `http` when the `SS_BASE_URL` is specified with `https`.

EG: `SS_BASE_URL="https://www.site.com"` will produce `<base href="http://www.site.com/"><!--[if lte IE 6]></base><![endif]-->` in the static files.

This causes "breakage" on the front-end when css & JS is blocked due to a downgraded protocol.

**Comment by chillu** [comment]
Right, it should honour `SS_BASE_URL` to avoid mixed content warnings - these won't appear when you load the error page over `http://`, but include CSS and JS via `https://`.

**Comment by tractorcow** [comment]
What if we used protocol-less absolute urls for resources? E.g. `<link href="http://somesite" />` becomes `<link href="//somesite">`. If your site is being served across multiple protocols, even if your static files contain `https` you'll get mixed content warnings if your page errors on a non-https url.

**Comment by axllent** [comment]
That's an interesting idea actually. I'm not sure whether that is (or isn't) allowed as I couldn't find an example in the specs of `//`.

**Comment by tractorcow** [comment]
Try it out in a project I think it works fine ;) You can just hard-code it into your template.

**Comment by axllent** [comment]
I don't think that the fact it works it too relevant, there is a LOT of broken stuff that works in browsers ;-) It's more as to whether it is allowed/intended...

> The base URL to be used throughout the document for relative URL addresses. If this attribute is specified, this element must come before any other elements with attributes whose values are URLs. Absolute and relative URLs are allowed.

Then the definition of [absolute URL](https://url.spec.whatwg.org/#syntax-url-absolute) states:

> An absolute-URL string must be one of the following
> 
> - a URL-scheme string that is an ASCII case-insensitive match for a special scheme and not an ASCII case-insensitive match for "file", followed by U+003A (:) and a scheme-relative-special-URL string
> - a URL-scheme string that is not an ASCII case-insensitive match for a special scheme, followed by U+003A (:) and a relative-URL string
> - a URL-scheme string that is an ASCII case-insensitive match for "file", followed by U+003A (:) and a scheme-relative-file-URL string

It's a rabbit hole and to be honest I cannot find any reference to, or example of a base url using the "protocol-less" variation. I suspect this may not only be bad practice, but also something that should be avoided for something like `base`. Surely though the base href protocol should/could be parsed from a defined `SS_BASE_URL`?

**Comment by axllent** [comment]
Sorry, I just found one of interest: https://stackoverflow.com/questions/21908318/can-i-use-a-protocol-relative-base-tag

**Comment by dhensby** [comment]
oh, I really wouldn't do protocol less for the base tag...

**Comment by tractorcow** [comment]
Could we statically cache both versions, and serve them up conditionally?

**Comment by tractorcow** [comment]
> The base URL to be used throughout the document for relative URL addresses. If this attribute is specified, this element must come before any other elements with attributes whose values are URLs. Absolute and relative URLs are allowed.

So it could be a relative url, just that google's crawler has a few issues with it.

I might suggest something that we can allow web devs to configure, rather than fix them to one solution.

**Comment by chillu** [comment]
I think it should take `SS_BASE_URL` into consideration. I don't think we need to get into any conditional serving, that's overkill. If you're running `https://` on your site, it's best practice to set up a blanket redirect anyway - in which case any 404s would be served from `https://`. I acknowledge that there's some situations where you need to run mixed (e.g. when third party `<script>` includes can't be served over `https://`), but that's a fringe case not worth spending a lot of time on.

**Comment by tractorcow** [comment]
I guess subsites needs separate error files with different bases?

**Comment by dhensby** [comment]
> No, I think that means that relative urls (eg <base href="/">) are fine

they definitely aren't - I've tried that before, it's not compliant... works in some browsers but it's not meant to

**Comment by chillu** [comment]
I'd put this back on `impact/high`, since I agree that mixed content warnings on "not found" and "server error" pages are bad. And this will affect most installs, since HTTPS is becoming fairly strongly encouraged by browsers now. 

But: Can't you just work around the issue by not including the base tag in your template in the first place? As far as I understand, the only reason it's there is to allow serving SilverStripe from a subfolder (like `http://localhost/site1` vs. `http://localhost/site2`). With the increasing use of VMs, I think that's an outdated concept, and we should remove support for it from SilverStripe 5.x.

**Comment by axllent** [comment]
That's a very valid point @chillu - and ironically it hadn't even crossed my mind (that I can recall) *not* to use the base tag, probably because it has always been a default is all templates including those still provided by SilverStripe.

I actually still use the `http://localhost/site2` format for a lot of my development and probably will continue to for a great number of development sites. Having said that, I don't _think_ the exclusion of the base tag should even affect those (at all) as we don't use the `my-page/123/` format and `$Link` generates relative (to base) links... unless it is tied into the CMS for any reason.

OK, point noted, I will start removing these unless absolutely necessary in the meantime. +1 for getting it resolved anyway, thank you!

**Comment by wernerkrauss** [comment]
This is still a thing in SS6.1. I found to work around this by adding:

```
---
Only:
  envvarset: 'SS_BASE_URL'
---
Silverstripe\Control\Director:
  alternate_base_url: '`SS_BASE_URL`'
```

(see https://github.com/silverstripe/silverstripe-framework/issues/10296#issuecomment-1115841052)

IMHO this should not be needed and it's a really bad bug to me, as browsers don't allow loading http resources in a https page any more.

**Comment by wernerkrauss** [comment]
This is a just workaround for single domain usage. When having e.g. a seperate domain for admin (with different caching options), those get redirected to the alternate_base_url...

**Comment by adiwidjaja** [comment]
Ok, the problem with the solution from @wernerkrauss is that you cannot have a secondary domain - I have one (with nginx fcgi cache) as a public url and one for the backend users.

For that, I found a solution: I have the yml like werner and then set the `SS_BASE_URL` just in the deployment (and other cli scripts):

`SS_BASE_URL="https://www.mysite.com/" vendor/bin/sake db:build --flush`

This way, the base_url is correct for cli scripts (404 page generation) and flexible for other requests.
